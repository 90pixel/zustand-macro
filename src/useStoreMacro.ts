import { addNamed as addImportDeclaration } from "@babel/helper-module-imports";
import { MacroError, MacroHandler } from "babel-plugin-macros";

type MacroConfig = {
  importName: string,
  importSource: string,
}

const useStoreMacro: MacroHandler = ({ references, state, babel, config }) => {
  const defaultImport = references.default || [];
  const t = babel.types;

  if (!config)
    throw new MacroError("Macro config must be provided in package.json");

  const macroConfig: MacroConfig = config.useStore;

  defaultImport.forEach(referencePath => {
    const functionCallPath = referencePath.parentPath;

    if (functionCallPath?.type !== "CallExpression")
      throw new MacroError("useStoreMacro only supports function calls.")

    addImportDeclaration(
      referencePath,
      macroConfig.importName,
      macroConfig.importSource,
      {
        nameHint: `_${macroConfig.importName}`
      });

    const idPath = functionCallPath.parentPath?.get("id");

    if (!t.isObjectPattern(idPath))
      throw new MacroError("useStoreMacro only supports object destructuring.");

    const deconstructedProps = (idPath as any).node
      .properties
      .map((x: any) => x.value.name);

    const useStoreParameters = deconstructedProps
      .map((x: any) => `${x}: s.${x}`)
      .join(",");

    const useShallowStoreNode = babel.template.expression.ast(`
      _${macroConfig.importName}((s) => ({
        ${useStoreParameters}
      }))
    `);

    functionCallPath?.replaceWith(useShallowStoreNode);
  });
}

export default useStoreMacro;
