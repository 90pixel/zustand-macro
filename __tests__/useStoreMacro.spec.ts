import pluginTester from "babel-plugin-tester";
import babelPluginMacros from "babel-plugin-macros";

pluginTester({
  plugin: babelPluginMacros,
  pluginName: "useStoreMacro",
  babelOptions: { filename: __filename },
  tests: {
    "generate actual shallow store implementation": {
      code: `
        import useStoreMacro from "../src/index.macro";

        const {
          count,
          decrementCount,
          incrementCount,
        } = useStoreMacro();
      `,
      snapshot: true,
    },
    "should fail if not a function call": {
      code: `
        import useStoreMacro from "../src/index.macro";

        const {
          count,
          decrementCount,
          incrementCount,
        } = useStoreMacro;
      `,
      error: "useStoreMacro only supports function calls.",
      snapshot: true,
    },
    "should support object destructuring assignment": {
      code: `
        import useStoreMacro from "../src/index.macro";

        const {
          count,
          decrementCount: decrement,
          incrementCount: increment,
        } = useStoreMacro();
      `,
      snapshot: true,
    },
  }
})
