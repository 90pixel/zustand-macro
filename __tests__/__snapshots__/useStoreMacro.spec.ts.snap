// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`useStoreMacro generate actual shallow store implementation: generate actual shallow store implementation 1`] = `

import useStoreMacro from "../src/index.macro";

const {
  count,
  decrementCount,
  incrementCount,
} = useStoreMacro();

      ↓ ↓ ↓ ↓ ↓ ↓

import { useShallowStore as _useShallowStore } from "hooks";

const { count, decrementCount, incrementCount } = _useShallowStore((s) => ({
  count: s.count,
  decrementCount: s.decrementCount,
  incrementCount: s.incrementCount,
}));


`;

exports[`useStoreMacro should fail if not a function call: should fail if not a function call 1`] = `

import useStoreMacro from "../src/index.macro";

const {
  count,
  decrementCount,
  incrementCount,
} = useStoreMacro;

      ↓ ↓ ↓ ↓ ↓ ↓

MacroError: /Users/serkan/repos/zustand-macro/__tests__/useStoreMacro.spec.ts: useStoreMacro only supports function calls.

`;

exports[`useStoreMacro should support object destructuring assignment: should support object destructuring assignment 1`] = `

import useStoreMacro from "../src/index.macro";

const {
  count,
  decrementCount: decrement,
  incrementCount: increment,
} = useStoreMacro();

      ↓ ↓ ↓ ↓ ↓ ↓

import { useShallowStore as _useShallowStore } from "hooks";

const {
  count,
  decrementCount: decrement,
  incrementCount: increment,
} = _useShallowStore((s) => ({
  count: s.count,
  decrementCount: s.decrementCount,
  incrementCount: s.incrementCount,
}));


`;
